media/src/iosMain/kotlin/DefaultMediaManager.kt

```diff
@@ -193,6 +221,59 @@ actual class DefaultMediaManager(mediaSurfaceProvider: MediaSurfaceProvider?, co
+        observers.value = listOf(
+            observeNotification(AVPlayerItemDidPlayToEndTimeNotification) {
+                avPlayer.currentItem?.let { currentItem ->
+                    // Completion event may come in after a rewind has occurred. To ensure it is intended to complete, we compare current time to duration
+                    // Current time may be slightly off however, so we maintain a margin of 1 second
+                    val secondsFromCompletion = CMTimeGetSeconds(CMTimeSubtract(currentItem.duration, currentItem.currentTime()))
+                    if (CMTimeCompare(currentItem.duration, kCMTimeIndefinite.readValue()) == 0 || secondsFromCompletion.absoluteValue < 1.0) {
+                        handleCompleted()
+                    } else {
+                        // Otherwise reset the rate since completion events will pause playback
+                        avPlayer.rate = avPlayer.defaultRate
+                    }
+                }
+            },
+            observeNotification(AVPlayerItemFailedToPlayToEndTimeNotification) {
+                (it?.userInfo?.get(AVPlayerItemFailedToPlayToEndTimeErrorKey) as? NSError)?.handleError()
+            },
+            observeNotification(AVPlayerRateDidChangeNotification) { notification ->
+                // When no reason is given, it is unknown. This includes completion and Audio Session Route changes, which are covered by other notifications
+                notification?.userInfo?.get(AVPlayerRateDidChangeReasonKey)?.let {
+                    handleRateChanged(avPlayer.rate)
+                }
+            },
+            observeNotification(UIApplicationDidEnterBackgroundNotification) {
+                surface?.bind?.invoke(null)
+            },
+            observeNotification(UIApplicationWillEnterForegroundNotification) {
+                surface?.bind?.invoke(avPlayer)
+            },
+            observeNotification(AVAudioSessionRouteChangeNotification) { notification ->
+                val reason = notification?.userInfo?.get(AVAudioSessionRouteChangeReasonKey) as Int
+                when (reason.toULong()) {
+                    AVAudioSessionRouteChangeReasonUnknown,
+                    AVAudioSessionRouteChangeReasonOldDeviceUnavailable,
+                    -> {
+                        // This Route change will cause rate to be set to 0.0
+                        // If we dont want that, we should wait until the change has occurred and then revert it
+                        if (settings.playAfterDeviceUnavailable) {
+                            launch {
+                                avPlayer.observeKeyValueAsFlow<Float>("rate").first { it == 0.0f }
+                                avPlayer.rate = avPlayer.defaultRate
+                            }
+                        } else {
+                            // This is an unknown reason for a rate change so we handle it here
+                            handleRateChanged(0.0f)
```

Comment:
I think it's more reliable to observe key value (rate from the player, similar to volume above) rather then guessing it here